packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2876,17 +2892,22 @@ export class ContainerRuntime
 			}
 		}
 
-		if (envelope.address === undefined) {
+		const address = envelope.address;
+		if (address === undefined) {
 			// No address indicates a container signal message.
 			this.emit("signal", transformed, local);
 			return;
 		}
 
+		if (this.handleSignal(address, transformed, local)) {
+			return;
+		}
+
 		// Due to a mismatch between different layers in terms of
 		// what is the interface of passing signals, we need to adjust
 		// the signal envelope before sending it to the datastores to be processed
 		const envelope2: IEnvelope = {
-			address: envelope.address,
+			address,
 			contents: transformed.content,
 		};
 		transformed.content = envelope2;
```

Comment:
Just below, we have this.channelCollection.processSignal().  Which is really just "read the envelope's address and forward to the correct recipient".

Maybe this would be a reasonable place to introduce your new routing, if it were generalized to something effectively like this.addressableRecipients.processSignal()?  I'm mostly thinking it would be nice if there's a single place that holds knowledge of valid addresses and the responsibility of routing to them.

Reply:
I've made an incremental step in that direction.
DataStore addressing and typing is fixed (not changable) and will remain exceptional.
And unless there is a resolution to what can be internal or not inheritance means logic will be separated.